spec/models/cart_product_spec.rb

```diff
@@ -0,0 +1,132 @@
+      expect(subject).to have_db_column(:cart_id).of_type(:integer).with_options(null: false)
+      expect(subject).to have_db_column(:product_id).of_type(:integer).with_options(null: false)
+      expect(subject).to have_db_column(:created_at).of_type(:datetime).with_options(null: false)
+      expect(subject).to have_db_column(:updated_at).of_type(:datetime).with_options(null: false)
+    end
+
+    it 'has indexes' do
+      expect(subject).to have_db_index(:cart_id)
+      expect(subject).to have_db_index(:product_id)
+    end
+  end
+
+  describe 'associations' do
+    it 'has relationships' do
+      expect(subject).to belong_to(:cart)
+      expect(subject).to belong_to(:product)
+    end
+  end
+
+  describe 'validations' do
+    context 'when is an attribute supervision' do
+      it 'validates the quantity attribute' do
+        is_expected.to validate_presence_of(:quantity)
+      end
+
+      it 'validates the total_amount attribute' do
+        is_expected.to validate_presence_of(:total_amount)
+      end
+
+      it 'validates the quantity attribute is greater or equal than 0' do
+        is_expected.to validate_numericality_of(:quantity).is_greater_than_or_equal_to(0)
+      end
+
+      it 'validates the total_amount attribute is greater or equal than 0' do
+        is_expected.to validate_numericality_of(:total_amount).is_greater_than_or_equal_to(0)
+      end
+    end
+
+    context 'when is a supervision with the product' do
+      let(:cart_product) { build(:cart_product, product: product, cart: cart, quantity: quantity) }
+      let(:cart) { create(:cart, :with_user) }
+      let(:quantity) { 5 }
+
+      context 'when is an active product' do
+        let(:product) { create(:product) }
+
+        it 'is a valid record' do
+          expect(cart_product.valid?).to be_truthy
+        end
+      end
+
+      context 'when is an inactive product' do
+        let(:product) { create(:product, status: 0) }
+
+        it 'raises an error' do
+          expect { cart_product.save! }.to raise_error(ActiveRecord::RecordInvalid)
+        end
+      end
+
+      context 'when product has enough stock' do
+        let(:product) { create(:product) }
+
+        it 'is a valid record' do
+          expect(cart_product.valid?).to be_truthy
+        end
+      end
+
+      context 'when product does not have enough stock' do
+        let(:product) { create(:product, stock: 1) }
+
+        it 'raises an error' do
+          expect { cart_product.save! }.to raise_error(ActiveRecord::RecordInvalid)
+        end
+      end
+    end
+  end
+
+  describe '#create' do
+    let(:cart_product) { build(:cart_product, product: product, cart: cart, quantity: quantity) }
+    let(:cart) { create(:cart, :with_user) }
+    let(:quantity) { 5 }
+
+    context 'when transaction is successful' do
+      let(:product) { create(:product) }
+
+      it 'is saved in database' do
+        expect { cart_product.save! }.to change { CartProduct.count }.by(1)
```

Comment:
What you mention is correct, but since this is a critical process, a kind of redundancy can be generated since when creating or updating a record, this has an impact on another associated model. This allows us to have a kind of extra checker for the process of adding or removing products to the cart.

I share the idea that it is not necessary to generate tests for features of the framework that are clearly already tested.